File: programming_exercise_2/R/ex2.R
# Copyright (C) 2014  Caleb Lo
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.

# Machine Learning
# Programming Exercise 2: Logistic Regression
# Problem: Predict chances of university admission for an applicant given data
# for admissions decisions and test scores of various applicants

PlotData <- function(X, y) {
  # Plots data.
  #
  # Args:
  #   X: Features to be plotted.
  #   y: Data labels.
  #
  # Returns:
  #   None.
  positiveIndices <- which(y == 1)
  negativeIndices <- which(y == 0)
  positiveExamples <- cbind(X[positiveIndices, ])
  negativeExamples <- cbind(X[negativeIndices, ])
  plot(positiveExamples[, 1], positiveExamples[, 2], cex=1.75, pch="+",
       xlab="", ylab="", xlim=c(30, 100), ylim=c(30, 100))
  points(x=negativeExamples[, 1], y=negativeExamples[, 2], col="yellow",
         bg="yellow", cex=1.75, pch=22, xlab="", ylab="")
  return(0)
}

PlotDecisionBoundary <- function(X, y, theta) {
  # Plots decision boundary.
  #
  # Args:
  #   X: Features that have already been plotted.
  #   y: Data labels.
  #   theta: Parameter that determines slope of decision boundary.
  #
  # Returns:
  #   None.
  returnCode <- PlotData(cbind(X[, 2], X[, 3]), y)
  yLineVals <- (theta[1] + theta[2] * X[, 2]) / (-1 * theta[3])
  lines(cbind(X[, 2]), yLineVals, col="blue", pch="-")
  return(0)
}

ReadKey <- function() {
  # Reads key press.
  #
  # Args:
  #   None.
  #
  # Returns:
  #   None.
  cat("Program paused.  Press enter to continue.")
  line <- readline()
  return(0)
}

ComputeSigmoid <- function(z) {
  # Computes sigmoid function.
  #
  # Args:
  #   z: Can be a scalar, a vector or a matrix.
  #
  # Returns:
  #   sigmoidZ: Sigmoid function value.
  sigmoidZ <- 1 / (1 + exp(-z))
  return(sigmoidZ)
}

ComputeCost <- function(theta, X, y, numTrainEx) {
  # Computes cost function J(\theta).
  #
  # Args:
  #   theta: Vector of parameters for logistic regression.
  #   X: Matrix of features.
  #   y: Vector of labels.
  #   numTrainEx: Number of training examples.
  #
  # Returns:
  #   jTheta: Logistic regression cost.
  if (numTrainEx > 0) {
    hTheta <- ComputeSigmoid(X %*% theta)
    jTheta <- (colSums(-y * log(hTheta) 
                       - (1 - y) * log(1 - hTheta))) / numTrainEx
  } else {
    stop('Insufficient training examples')
  }
  return(jTheta)  
}

ComputeGradient <- function(theta, X, y, numTrainEx) {
  # Computes gradient of cost function J(\theta).
  #
  # Args:
  #   theta: Vector of parameters for logistic regression.
  #   X: Matrix of features.
  #   y: Vector of labels.
  #   numTrainEx: Number of training examples.
  #
  # Returns:
  #   gradArray: Vector of logistic regression gradients (one per feature).
  numFeatures <- dim(X)[2]
  if (numFeatures > 0) {
    if (numTrainEx > 0) {
      hTheta <- ComputeSigmoid(X %*% theta)
      gradArray <- matrix(0, numFeatures, 1)
      gradTermArray <- matrix(0, numTrainEx, numFeatures)
      for (gradIndex in 1:numFeatures) {
        gradTermArray[, gradIndex] <- (hTheta - y) * X[, gradIndex]
        gradArray[gradIndex] <- 
          (sum(gradTermArray[, gradIndex])) / (numTrainEx)
      }
    } else {
      stop('Insufficient training examples')
    }
  } else {
    stop('Insufficient features')
  }
  return(gradArray)
}

ComputeCostGradList <- function(X, y, theta) {
  # Aggregates computed cost and gradient.
  #
  # Args:
  #   X: Matrix of features.
  #   y: Vector of labels.
  #   theta: Vector of parameters for logistic regression.
  #
  # Returns:
  #   returnList: List of two objects.
  #               jTheta: Updated vector of parameters for logistic regression.
  #               gradArray: Updated vector of logistic regression gradients 
  #                          (one per feature).
  numTrainEx <- dim(y)[1]
  jTheta <- ComputeCost(theta, X, y, numTrainEx)
  gradArray <- ComputeGradient(theta, X, y, numTrainEx)
  returnList <- list("jTheta"=jTheta, "gradArray"=gradArray)
  return(returnList)
}

LabelPrediction <- function(X, theta) {
  # Performs label prediction on training data.
  #
  # Args:
  #   X: Matrix of features.
  #   theta: Vector of parameters for logistic regression.
  #
  # Returns:
  #   p: Vector of predictions (one per example).
  sigmoidArr <- ComputeSigmoid(X %*% theta)
  p <- (sigmoidArr >= 0.5)
  return(p)
}

# Use setwd() to set working directory to directory that contains this source
# file
# Load file into R
applicantData <- read.csv("../applicantData.txt", header=FALSE)

# Plot data
print("Plotting data with + indicating (y = 1) examples and o indicating 
      (y = 0) examples.")
returnCode <- PlotData(cbind(applicantData[, "V1"], applicantData[, "V2"]),
                       applicantData[, "V3"])
title(xlab="Exam 1 score", ylab="Exam 2 score")
plotLegend <- legend('bottomright', col=c("black", "yellow"), 
                     pt.bg=c("black", "yellow"), pch=c(43, 22), pt.cex=1.75, 
                     legend=c("", ""), bty="n", trace=TRUE)
text(plotLegend$text$x - 3, plotLegend$text$y, c('Admitted', 'Not admitted'),
     pos=2)
returnCode <- ReadKey()
numTrainEx <- dim(applicantData)[1]
numFeatures <- dim(applicantData)[2] - 1
onesVec <- t(t(rep(1, numTrainEx)))
xMat <- cbind(onesVec, applicantData[, "V1"], applicantData[, "V2"])
yVec <- cbind(applicantData[, "V3"])
thetaVec <- t(t(rep(0, numFeatures + 1)))

# Compute initial cost and gradient
initComputeCostList <- ComputeCostGradList(xMat, yVec, thetaVec)
print(sprintf("Cost at initial theta (zeros): %.6f", 
              initComputeCostList$jTheta))
print(sprintf("Gradient at initial theta (zeros): "))
cat(format(round(initComputeCostList$gradArray, 6), nsmall=6), sep="\n")
returnCode <- ReadKey()

# Use optim to solve for optimum theta and cost
optimResult <- optim(thetaVec, fn=ComputeCost, gr=ComputeGradient, xMat, yVec,
                     numTrainEx, method="BFGS", control=list(maxit=400))
print(sprintf("Cost at theta found by optim: %.6f", optimResult$value))
print(sprintf("theta: "))
cat(format(round(optimResult$par, 6), nsmall=6), sep="\n")
returnCode <- PlotDecisionBoundary(xMat, yVec, optimResult$par)
title(xlab="Exam 1 score", ylab="Exam 2 score")
plotLegend <- legend('bottomright', col=c("black", "yellow", "blue"),
                     pt.bg=c("black", "yellow", "blue"), pch=c(43, 22, 45),
                     pt.cex=1.75, legend=c("", "", ""), bty="n", trace=TRUE)
text(plotLegend$text$x - 3, plotLegend$text$y, 
     c('Admitted', 'Not admitted', 'Decision Boundary'), pos=2)
returnCode <- ReadKey()

# Predict admission probability for a student with score 45 on exam 1 and score
# 85 on exam 2
admissionProb <- ComputeSigmoid(cbind(1, 45, 85) %*% optimResult$par)
print(sprintf("For a student with scores 45 and 85, we predict an admission 
              probability of %.6f", admissionProb))

# Compute accuracy on training set
trainingPredict <- (LabelPrediction(xMat, optimResult$par) + 0)
print(sprintf("Train Accuracy: %.6f",
              100 * apply((trainingPredict == yVec), 2, mean)))
returnCode <- ReadKey()

File: programming_exercise_2/Python/ex2.py
# Copyright (C) 2014  Caleb Lo
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.

# Machine Learning
# Programming Exercise 2: Logistic Regression
# Problem: Predict chances of university admission for an applicant given data
# for admissions decisions and test scores of various applicants
from matplotlib import pyplot
from scipy.optimize import fmin_ncg
import numpy

class Error(Exception):
    def __init__(self, value):
        self.value = value
    def __str__(self):
        return repr(self.value)


def plot_data(x, y):
    """ Plots data.

    Args:
      x: Features to be plotted.
      y: Data labels.

    Returns:
      None.
    """
    positive_indices = numpy.where(y == 1)
    negative_indices = numpy.where(y == 0)
    pos = pyplot.scatter(x[positive_indices, 0], x[positive_indices, 1], s=80,
                         marker='+', color='k')
    pyplot.hold(True)
    neg = pyplot.scatter(x[negative_indices, 0], x[negative_indices, 1], s=80,
                         marker='s', color='y')
    pyplot.hold(False)
    pyplot.ylabel('Exam 2 score', fontsize=18)
    pyplot.xlabel('Exam 1 score', fontsize=18)
    return None


def plot_decision_boundary(x, y, theta):
    """ Plots decision boundary.

    Args:
      x: Features that have already been plotted.
      y: Data labels.
      theta: Parameter that determines slope of decision boundary.

    Returns:
      None.
    """
    plot_data(numpy.c_[x[:, 1], x[:, 2]], y)
    pyplot.hold(True)
    y_line_vals = (theta[0]+theta[1]*x[:, 1])/(-1*theta[2])
    pyplot.plot(x[:, 1], y_line_vals, 'b-', markersize=18)
    pyplot.hold(False)
    return None


def compute_sigmoid(z):
    """ Computes sigmoid function.

    Args:
      z: Can be a scalar, a vector or a matrix.

    Returns:
      sigmoid_z: Sigmoid function value.
    """
    sigmoid_z = 1/(1+numpy.exp(-z))
    return sigmoid_z


def compute_cost(theta, X, y, num_train_ex):
    """ Computes cost function J(\theta).

    Args:
      theta: Vector of parameters for logistic regression.
      X: Matrix of features.
      y: Vector of labels.
      num_train_ex: Number of training examples.

    Returns:
      j_theta: Logistic regression cost.

    Raises:
      An error occurs if the number of features is 0.
      An error occurs if the number of training examples is 0.
    """
    if (num_train_ex == 0): raise Error('num_train_ex = 0')
    num_features = X.shape[1]
    if num_features == 0: raise Error('num_features = 0')
    theta = numpy.reshape(theta, (num_features, 1), order='F')
    h_theta = compute_sigmoid(numpy.dot(X, theta))
    j_theta = (numpy.sum(numpy.subtract(numpy.multiply(-y, numpy.log(h_theta)),
                                        numpy.multiply((1-y),
                                                       numpy.log(1-h_theta))),
                         axis=0))/num_train_ex
    return j_theta


def compute_gradient(theta, X, y, num_train_ex):
    """ Computes gradient of cost function J(\theta).

    Args:
      theta: Vector of parameters for logistic regression.
      X: Matrix of features.
      y: Vector of labels.
      num_train_ex: Number of training examples.

    Returns:
      grad_array_flat: Vector of logistic regression gradients
                       (one per feature).

    Raises:
      An error occurs if the number of features is 0.
      An error occurs if the number of training examples is 0.
    """
    if (num_train_ex == 0): raise Error('num_train_ex = 0')
    num_features = X.shape[1]
    if num_features == 0: raise Error('num_features = 0')
    theta = numpy.reshape(theta, (num_features, 1), order='F')
    h_theta = compute_sigmoid(numpy.dot(X, theta))
    grad_array = numpy.zeros((num_features, 1))
    for grad_index in range(0, num_features):
        grad_term = numpy.multiply(numpy.reshape(X[:, grad_index],
                                                 (num_train_ex, 1)),
                                   numpy.subtract(h_theta, y))
        grad_array[grad_index] = (numpy.sum(grad_term, axis=0))/num_train_ex
    grad_array_flat = numpy.ndarray.flatten(grad_array)
    return grad_array_flat


def compute_cost_grad_list(theta, X, y):
    """ Aggregates computed cost and gradient.

    Args:
      theta: Vector of parameters for logistic regression.
      X: Matrix of features.
      y: Vector of labels.


    Returns:
      return_list: List of two objects.
                   j_theta: Updated vector of parameters for logistic
                            regression.
                   grad_array: Updated vector of logistic regression gradients 
                               (one per feature).
    """
    num_features = X.shape[1]
    num_train_ex = y.shape[0]
    j_theta = compute_cost(theta, X, y, num_train_ex)
    grad_array_flat = compute_gradient(theta, X, y, num_train_ex)
    grad_array = numpy.reshape(grad_array_flat, (num_features, 1), order='F')
    return_list = {'j_theta': j_theta, 'grad_array': grad_array}
    return return_list


def label_prediction(X, theta):
    """ Performs label prediction on training data.

    Args:
      X: Matrix of features.
      theta: Vector of parameters for logistic regression.

    Returns:
      p: Vector of predictions (one per example).

    Raises:
      An error occurs if the number of training examples is 0.
    """
    num_train_ex = X.shape[0]
    if (num_train_ex == 0): raise Error('num_train_ex = 0')
    sigmoid_array = compute_sigmoid(numpy.dot(X, theta))
    p = numpy.zeros((num_train_ex, 1))
    for train_index in range(0, num_train_ex):
        if (sigmoid_array[train_index] >= 0.5):
            p[train_index] = 1
        else:
            p[train_index] = 0
    return p


def main():
    """ Main function

    Raises:
      An error occurs if the number of training examples is 0.
    """
    print("Plotting data with + indicating (y = 1) examples and o indicating (y = 0) examples.")
    print("")
    applicant_data = numpy.genfromtxt("../applicantData.txt", delimiter=",")
    return_code = plot_data(numpy.c_[applicant_data[:, 0],
                                     applicant_data[:, 1]],
                            applicant_data[:, 2])
    pyplot.legend(('Admitted', 'Not admitted'), loc='lower right')
    pyplot.show()
    input("Program paused. Press enter to continue.")
    print("")
    num_train_ex = applicant_data.shape[0]
    if (num_train_ex == 0): raise Error('num_train_ex = 0')
    num_features = applicant_data.shape[1]-1
    ones_vec = numpy.ones((num_train_ex, 1))
    x_mat = numpy.c_[applicant_data[:, 0], applicant_data[:, 1]]
    x_mat_aug = numpy.c_[ones_vec, x_mat]
    y_vec = numpy.reshape(applicant_data[:, 2], (num_train_ex, 1))
    theta_vec = numpy.zeros((num_features+1, 1))

    # Compute initial cost and gradient
    init_compute_cost_list = compute_cost_grad_list(theta_vec, x_mat_aug, y_vec)
    print("Cost at initial theta (zeros): %.6f" %
          init_compute_cost_list['j_theta'])
    print("Gradient at initial theta (zeros):")
    print("%s\n" %
          numpy.array_str(numpy.round(init_compute_cost_list['grad_array'], 6)))
    input("Program paused. Press enter to continue.")

    # Use fmin_ncg to solve for optimum theta and cost
    theta_vec_flat = numpy.ndarray.flatten(theta_vec)
    f_min_ncg_out = fmin_ncg(compute_cost, theta_vec_flat,
                             fprime=compute_gradient, args=(x_mat_aug, y_vec,
                                                            num_train_ex),
                             avextol=1e-10, epsilon=1e-10, maxiter=400,
                             full_output=1)
    theta_opt = numpy.reshape(f_min_ncg_out[0], (num_features+1, 1), order='F')
    print("Cost at theta found by fmin_ncg: %.6f" % f_min_ncg_out[1])
    print("theta:")
    print("%s\n" % numpy.array_str(numpy.round(theta_opt, 6)))
    return_code = plot_decision_boundary(x_mat_aug, y_vec, theta_opt)
    pyplot.legend(('Decision Boundary', 'Admitted', 'Not admitted'),
                  loc='lower left')
    pyplot.show()
    input("Program paused. Press enter to continue.")

    # Predict admission probability for a student with score 45 on exam 1 and
    # score 85 on exam 2
    admission_prob = compute_sigmoid(numpy.dot(numpy.array([1, 45, 85]),
                                               theta_opt))
    print("For a student with scores 45 and 85, we predict an admission probability of %.6f" %
          admission_prob)

    # Compute accuracy on training set
    training_predict = label_prediction(x_mat_aug, theta_opt)
    num_train_match = 0
    for train_index in range(0, num_train_ex):
        if (training_predict[train_index] == y_vec[train_index]):
            num_train_match += 1
    print("Train Accuracy: %.6f" % (100*num_train_match/num_train_ex))
    input("Program paused. Press enter to continue.")

# Call main function
if __name__ == "__main__":
    main()
